src/Dictionaries/ExecutablePoolDictionarySource.cpp

```diff
@@ -0,0 +1,272 @@
+{
+    LOG_TRACE(log, "loadIds {} size = {}", toString(), ids.size());
+
+    auto block = blockForIds(dict_struct, ids);
+    return getStreamForBlock(block);
+}
+
+BlockInputStreamPtr ExecutablePoolDictionarySource::loadKeys(const Columns & key_columns, const std::vector<size_t> & requested_rows)
+{
+    LOG_TRACE(log, "loadKeys {} size = {}", toString(), requested_rows.size());
+
+    auto block = blockForKeys(dict_struct, key_columns, requested_rows);
+    return getStreamForBlock(block);
+}
+
+BlockInputStreamPtr ExecutablePoolDictionarySource::getStreamForBlock(const Block & block)
+{
+    std::unique_ptr<ShellCommand> process;
+    process_pool->pop(process);
+
+    size_t rows_to_read = block.rows();
+    auto read_stream = context.getInputFormat(format, process->out, sample_block, rows_to_read);
+
+    auto stream = std::make_unique<PoolBlockInputStreamWithBackgroundThread>(
+        process_pool, std::move(read_stream), std::move(process), rows_to_read,
+        [block, this](WriteBufferFromFile & out) mutable
+        {
+            auto output_stream = context.getOutputStream(format, out, block.cloneEmpty());
+            formatBlock(output_stream, block);
+            std::cerr << "Write block to process " << std::endl;
+        });
+
+    if (implicit_key)
+        return std::make_shared<BlockInputStreamWithAdditionalColumns>(block, std::move(stream));
+    else
+        return std::shared_ptr<PoolBlockInputStreamWithBackgroundThread>(stream.release());
+}
+
+bool ExecutablePoolDictionarySource::isModified() const
+{
+    return true;
+}
+
+bool ExecutablePoolDictionarySource::supportsSelectiveLoad() const
+{
+    return true;
+}
+
+bool ExecutablePoolDictionarySource::hasUpdateField() const
+{
+    return !update_field.empty();
+}
+
+DictionarySourcePtr ExecutablePoolDictionarySource::clone() const
+{
+    return std::make_unique<ExecutablePoolDictionarySource>(*this);
+}
+
+std::string ExecutablePoolDictionarySource::toString() const
+{
+    return "ExecutablePool size: " + std::to_string(pool_size) + " command: " + command;
+}
+
+void registerDictionarySourceExecutablePool(DictionarySourceFactory & factory)
+{
+    auto create_table_source = [=](const DictionaryStructure & dict_struct,
+                                 const Poco::Util::AbstractConfiguration & config,
+                                 const std::string & config_prefix,
+                                 Block & sample_block,
+                                 const Context & context,
+                                 const std::string & /* default_database */,
+                                 bool check_config) -> DictionarySourcePtr
+    {
+        if (dict_struct.has_expressions)
+            throw Exception{"Dictionary source of type `executable_pool` does not support attribute expressions", ErrorCodes::LOGICAL_ERROR};
+
+        /// Executable dictionaries may execute arbitrary commands.
```

Comment:
But this commands will be run with the same privileges as clickhouse process?

Reply:
Yes.

Reply:
Yes, so they are dangerous. But it's impossible to create executable dictionaries without having root access to the server.

Maybe we will drop priveleges by `seccomp` and change to another user.
It will be addressed when we start developing UDFs.

Reply:
We have similar issue with `clickhouse-odbc-bridge` and we looking for various ways to drop privileges...